joplin/base/models/janis_page.py

```diff
@@ -43,38 +43,84 @@ class JanisBasePage(Page):
-        # TODO: add something to make this clear to users
-        return "#"
+        """
+        This function parses various attributes of related content types to construct the
+        expected url structure for janis.
+
+        For attributes with multiple relations, it ONLY takes the FIRST one.
+        """
+
+        try:
+            """
+             These use ternary operators with some appropriate conditionals
+             the idea is: return this value in these cases or tell use you got
+             nothing (see the privacy policy info page for example).
+
+             'None' responses get filtered out and removed from the URL path.
+
+             TODO:
+             make this more abstract(potentially not by content type)
+             further check if the order of conditionals affects performance
+             Better utilization of querysets may be possible for better performance
+            """
+            page_slug = self.slug or None
+            has_no_theme = [
+                'service page',
+                'topic page',
+                'information page',
+                'department page'
+            ]
+            has_no_topic_collection = has_no_theme
+
+            has_no_topic = [
+                'topic page',
+                'topic collection page',
+                'department page'
+            ]
+
+            theme_slug = (
+                self.theme.slug
+                if self.content_type.name not in has_no_theme
+                else None
+            )
+            # https://docs.djangoproject.com/en/2.2/ref/models/querysets/#first
+            topic_collection_slug = (
+                self.topiccollections.first().topiccollection.slug
+                if
+                (
+                    self.content_type.name not in has_no_topic_collection and
+                    # https://docs.djangoproject.com/en/2.2/ref/models/querysets/#exists
+                    self.topiccollections.exists()
+                )
+                else None
+            )
+            topic_slug = (
+                self.topics.first().topic.slug
+                if
+                (
+                    self.content_type.name not in has_no_topic and
+                    self.topics.exists()
+                )
+                else None
+            )
+
+            # add hardcoded language path to base url
+            base_url = os.environ["JANIS_URL"] + '/en'
+            # collect all our path elements
+            paths_list = [
+                base_url,
+                theme_slug,
+                topic_collection_slug,
+                topic_slug,
+                page_slug]
+            # join them together, filtering out empty ones
+            janis_url = '/'.join(filter(None, (paths_list)))
+            return janis_url
+        except Exception as e:
+            # right now this is a catch-all,
+            print("!janis url error!:", self.title, e)
+            return "#"
+            pass
```

Comment:
I like this solution a lot. I'm just wondering why you'd think an Exception would be thrown here? If anything, it looks like your solution is so complete, that we'd want the staging site to break and an Exception to be thrown if something catastrophically went wrong.

Reply:
Alternatively, if speed is a concern, should every content type have their own function to handle the construction of the janis_url? Then you could cut out a bunch of the conditional checks. Food for thought, not a mandate.

Reply:
Thanks! These are good questions. 

1) Regarding staging, I'm 90% sure this function exists merely to provide correct preview links for janis from joplin. 

2) As written it does cover a variety of edge cases, so the catch-all is more of a guarantee and a place to start if a new edge case is found. Previously these edge cases (like privacy policy) or guides would essentially not work but not give any obvious indication of that. But the code was not safe from bugs either, as Desi found. 

3) The way Python evaluates try/except blocks means that there is no performance downside to having this here, but it does leave a (albeit kinda hacky) message in the log before passing. So I'm not sure there is much of a downside to keeping it. (The Exception class, btw, does ignore some system and user input exceptions so its pretty safe to use here: https://docs.python.org/3/library/exceptions.html#exception-hierarchy )